api/src/test/java/org/openmrs/api/AdministrationServiceTest.java

```diff
@@ -168,119 +194,127 @@ public void setImplementationId_shouldThrowAPIExceptionIfGivenAPipeInTheImplemen
 		ImplementationId afterInvalid2 = adminService.getImplementationId();
 		assertNull("There shouldn't be an impl id defined after setting an invalid impl id", afterInvalid2);
 	}
-	
+
 	/**
 	 * @see AdministrationService#setImplementationId(ImplementationId)
 	 */
 	@Test
-	@Ignore
 	public void setImplementationId_shouldCreateImplementationIdInDatabase() {
 		// save a valid impl id
+		executeDataSet("org/openmrs/api/include/AdministrationServiceTest-general.xml");
 		ImplementationId validId = new ImplementationId();
-		validId.setImplementationId("JUNIT-TEST");
-		validId.setName("JUNIT-TEST implementation id");
-		validId.setPassphrase("This is the junit test passphrase");
+		validId.setImplementationId("JUNIT-TEST-3");
+		validId.setName("JUNIT-TEST-3 implementation id");
+		validId.setPassphrase("This is a junit test passphrase");
 		validId.setDescription("This is the junit impl id used for testing of the openmrs API only.");
-		adminService.setImplementationId(validId);
-		
-		assertEquals(validId, adminService.getImplementationId());
+		try {
+			Context.getAdministrationService().setImplementationId(validId);
+			assertNotNull(Context.getAdministrationService().getImplementationId());
+		}
+		catch (APIException e) {
+			//expected exception
```

Comment:
if you expect an exception to happen either put the exception in `@Test(expected=APIException.class)` or if this specific expection is thrown multiple times in the class and you cannot ensure the specific case with this style than use https://github.com/junit-team/junit4/wiki/Rules#expectedexception-rules where you can also assert on the exception message.

Reply:
Hey @teleivo , I have got a problem here. I think when running the tests when I am calling the function setImplementationId() , You can see that this function internally calls another function checkImplementationIdValidity() , see [HERE](https://github.com/openmrs/openmrs-core/blob/6966e508c199e3b808718d155373fa695a961ee9/api/src/main/java/org/openmrs/api/impl/AdministrationServiceImpl.java#L399) !! 

And this function always fails resulting in the execution of the catch block. Now I am not actually able to find out why the function checkImplementationIdValidity() is failing. Can you please help??

Reply:
I think thats why the try block below the function call remains uncovered :)

Reply:
this test to me seems to be the "happy path" so when setting the implementation id actually works, however there is an exception thrown my guess is because the id is not valid according to `checkImplementationIdValidity`. What is the exception?

the solution is not to catch it but to fix whats wrong in setting the implementation id.

the other tests for setImplementation id are not a good example, they should be changed to using
https://github.com/junit-team/junit4/wiki/Rules#expectedexception-rules an example of where we us it is here
https://github.com/openmrs/openmrs-core/blob/master/api/src/test/java/org/openmrs/OrderTest.java#L524-L526

Reply:
thank you @teleivo ! :) will do the changes.